xarray/core/concat.py

```diff
@@ -646,29 +656,37 @@ def get_indexes(name):
             # preserves original variable order
             result_vars[name] = result_vars.pop(name)
 
-    result = type(datasets[0])(result_vars, attrs=result_attrs)
-
-    absent_coord_names = coord_names - set(result.variables)
+    absent_coord_names = coord_names - set(result_vars)
     if absent_coord_names:
         raise ValueError(
             f"Variables {absent_coord_names!r} are coordinates in some datasets but not others."
         )
-    result = result.set_coords(coord_names)
-    result.encoding = result_encoding
 
-    result = result.drop_vars(unlabeled_dims, errors="ignore")
+    result_data_vars = {}
+    coord_vars = {}
+    for name, result_var in result_vars.items():
+        if name in coord_names:
+            coord_vars[name] = result_var
+        else:
+            result_data_vars[name] = result_var
 
     if index is not None:
-        # add concat index / coordinate last to ensure that its in the final Dataset
         if dim_var is not None:
             index_vars = index.create_variables({dim: dim_var})
         else:
             index_vars = index.create_variables()
-        result[dim] = index_vars[dim]
+
+        coord_vars[dim] = index_vars[dim]
         result_indexes[dim] = index
+        unlabeled_dims = unlabeled_dims - set([dim])
 
     # TODO: add indexes at Dataset creation (when it is supported)
-    result = result._overwrite_indexes(result_indexes)
+    coords_obj = Coordinates(coord_vars, indexes=result_indexes)
+
+    result = type(datasets[0])(result_data_vars, coords=coords_obj, attrs=result_attrs)
+    result.encoding = result_encoding
+
+    result = result.drop_vars(unlabeled_dims, errors="ignore")
```

Comment:
This smells a bit, especially with `errors="ignore"`. Remove these from `coord_vars` instead?

Why is it needed? At the very least we should add a cleanup TODO.

Reply:
Took a while to find but I managed to fix this in https://github.com/pydata/xarray/pull/8872/commits/b0e36124c04255378ee3f7928fb5d0276f628f89 🙂 

Should be a performance improvement too because now we don't create new dimension coordinates just to drop them at the end 🤯

Reply:
(Aside - This is weird imo)

```python
In [2]: import xarray as xr

In [3]: ds = xr.Dataset(data_vars={"a": ("x", [0])})

In [4]: 'x' in ds.coords
Out[4]: False

In [5]: ds.coords['x']
Out[5]: 
<xarray.DataArray 'x' (x: 1)> Size: 8B
array([0])
Dimensions without coordinates: x
```